packages/block-editor/src/components/media-replace-flow/index.js

```diff
@@ -0,0 +1,203 @@
+/**
+ * WordPress dependencies
+ */
+import { useState, createRef } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { speak } from '@wordpress/a11y';
+import {
+	FormFileUpload,
+	NavigableMenu,
+	MenuItem,
+	Toolbar,
+	Button,
+	Popover,
+	withNotices,
+} from '@wordpress/components';
+import {
+	LEFT,
+	RIGHT,
+	UP,
+	DOWN,
+	BACKSPACE,
+	ENTER,
+} from '@wordpress/keycodes';
+import { useSelect } from '@wordpress/data';
+import { compose } from '@wordpress/compose';
+
+/**
+ * Internal dependencies
+ */
+import MediaUpload from '../media-upload';
+import MediaUploadCheck from '../media-upload/check';
+import LinkEditor from '../url-popover/link-editor';
+import LinkViewer from '../url-popover/link-viewer';
+
+const MediaReplaceFlow = (
+	{
+		mediaURL,
+		allowedTypes,
+		accept,
+		onSelect,
+		onSelectURL,
+		onError,
+		name = __( 'Replace' ),
+	}
+) => {
+	const [ showURLInput, setShowURLInput ] = useState( false );
+	const [ showEditURLInput, setShowEditURLInput ] = useState( false );
+	const [ mediaURLValue, setMediaURLValue ] = useState( mediaURL );
+	const [ showMediaReplaceOptions, setShowMediaReplaceOptions ] = useState( false );
+	const mediaUpload = useSelect( ( select ) => {
+		return select( 'core/block-editor' ).getSettings().mediaUpload;
+	} );
+	const editMediaButtonRef = createRef();
+
+	const stopPropagation = ( event ) => {
+		event.stopPropagation();
+	};
+
+	const stopPropagationRelevantKeys = ( event ) => {
+		if ( [ LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER ].indexOf( event.keyCode ) > -1 ) {
+			// Stop the key event from propagating up to ObserveTyping.startTypingInTextField.
+			event.stopPropagation();
+		}
+	};
+
+	const selectMedia = ( media ) => {
+		onSelect( media );
+		setMediaURLValue( media.url );
```

Comment:
I can't seem to reproduce it reliably, but I encountered an error occurring at this line, when using "Replace" -> "Upload". The image got stuck in the spinning, uploading state. Are there cases that `media` could be `undefined` here that we're not accounting for?

```
index.js:49 Uncaught (in promise) TypeError: Cannot read property 'url' of undefined
    at selectMedia (index.js:49)
    at setMedia (index.js:61)
    at setAndUpdateFiles (upload-media.js:80)
    at _callee$ (upload-media.js:183)
    at tryCatch (wp-polyfill.js?ver=7.4.4:6496)
    at Generator.invoke [as _invoke] (wp-polyfill.js?ver=7.4.4:6722)
    at Generator.prototype.<computed> [as throw] (wp-polyfill.js?ver=7.4.4:6548)
    at asyncGeneratorStep (asyncToGenerator.js:3)
    at _throw (asyncToGenerator.js:29)
```

Reply:
Tried to run into this by:

- uploading small images
- uploading too large images
- uploading many images in a row

No go! :) i could simply add a check though.

Reply:
> i could simply add a check though.

If there is an issue, I'd personally rather that it fail loudly so that we can know that it exists, rather than silently absorb it via a condition (which might have other unintended side-effects or leave related issues without an obvious means to know that they exist).

Since I'm not sure how exactly I managed to get into that problem state, it may well not be a real issue but rather something about the environment I was running in. Otherwise, the only other thought I'd have is to explore which logic flows in the code could _ever possibly_ result in an `undefined` value here.